middle_end/flambda2.0/to_cmm/un_cps.ml

```diff
@@ -555,6 +556,11 @@ let machtype_of_kinded_parameter p =
   machtype_of_kind (Kinded_parameter.kind p)
 
 let machtype_of_return_arity = function
+  (* Functions that never return have arity 0. In that case, we
+     use the most restrictive machtype to ensure that the return
+     value of the function is not used. *)
+  | [] -> typ_addr
```

Comment:
It would seem more natural to use `Cmm.typ_void`.

Reply:
Right, but I'm worried that it would likely break this line of selectgen : https://github.com/ocaml/ocaml/blob/d4dea7afea89f3e6719b7148d5a659fbd62c529d/asmcomp/selectgen.ml#L170

I.e. using typ_void vould make one branch of the bound check conditional return a `[| Val |]`, and the other return `[| |]` (though I haven't tested it). Actually, that makes me wonder what happens in the case of flat float arrays (the soundness would depend on where the unboxing occurs on reads, Ill check tomorrow).

Reply:
For the branches of the conditional, that should be fine, as they won't be merged together (that's the whole point of not giving a return continuation to the C call). The problem is that you would like `Selectgen.emit_expr` to return `None` instead of `Some [| |]`, so that it correctly understands that this function never returns (like `raise`).
This is not possible without a patch to Cmm, though.

Reply:
Should have been fixed by the recent commits